lib/tokenizer/preprocessor.js

```diff
@@ -26,15 +31,34 @@ var Preprocessor = module.exports = function () {
 
     this.pos = -1;
     this.lastGapPos = -1;
+    this.lastCharPos = -1;
+    this.droppedBufferSize = 0;
 
     this.gapStack = [];
 
     this.skipNextNewLine = false;
 
     this.lastChunkWritten = false;
     this.endOfChunkHit = false;
+    this.bufferWaterline = DEFAULT_BUFFER_WATERLINE;
 };
 
+Object.defineProperty(Preprocessor.prototype, 'sourcePos', {
+    get: function () {
+        return this.droppedBufferSize + this.pos;
+    }
+});
+
+Preprocessor.prototype.dropParsedChunk = function () {
+    if (this.pos > this.bufferWaterline) {
```

Comment:
> Just trying to think of the cases where we will still have congestion inside of the streaming parser when chunks were already fed to it, but it didn't output the text node (because it was all text).

We will not have such issues with text content, but we will have such issue e.g. with extremely big attribute names/values. In that case our mechanism will not work and we'll buffer huge chunk of data.

> Is bufferWaterline supposed to be configurable via options btw?

I'm not sure if we should. It doesn't pass basic assertion for public API: how you will describe it to users? Maximum amount of data buffered by parser? It's not true, because we don't buffer exactly this amount and also we have edge cases there we can potentially buffer much more than that limit. I believe we should keep it private, since I would like to leave a space for improvement maybe we'll drastically modify this mechanism to handle edge cases.